### PR TITLE
Convert persist_source to a regular non-async operator

### PR DESCRIPTION
**Pull request by antiguru**
Convert the `persist_source` `decode_and_mfp` operator to a plain Timely operator without using `async`. This avoids maintaining Timely state across yields and seems simpler to reason about in general.

This also takes the intended changes from #17156 to compact the output in-place using the `ConsolidateBuffer`. This way, we don't have to worry about large allocations or spending significant time in consolidating updates.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a
  companion cloud PR to account for those changes that is tagged with
  the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by danhhz** [comment]
I'm supportive of the overall direction, but the original behavior of `yield_fn` was a suggestion from @frankmcsherry, so I'd love to get signoff from him on the change in behavior before looking at the rest of this too closely

File: src/storage-client/src/source/persist_source.rs
Comment (teskje) [inline src/storage-client/src/source/persist_source.rs:139]:
```suggestion
/// Pending work to read from fetched parts
```

**Review by vmarcos — COMMENTED** [review]
I will defer to storage folks regarding the details of the operator code; at least upon a comparison of the old and new versions, it seems to me that the logic has been kept unharmed.

Regarding the changes to `ConsolidateBuffer`, if we now introduce the function `give_at`, would it make sense to have a variation of `new` that does not require the output port to be specified, with an explicit `Option` being passed in? I am a bit worried of misuses where one gives a `port` in a multi-output operator, but calls `give_at` by mistake. With some extra checking, the implementation could potentially catch these problems?

**Comment by antiguru** [comment]
I updated the PR to use the `yield_fn` which was there previously. The PR is now best viewed with whitespace hidden!

File: src/storage-client/src/source/persist_source.rs
Comment (danhhz) [inline src/storage-client/src/source/persist_source.rs:226]: nit: this still references `fuel` but we switched back to `yield_fn`

File: src/storage-client/src/source/persist_source.rs
Comment (danhhz) [inline src/storage-client/src/source/persist_source.rs:175]: super nit: `TODO` comments are common enough that this variable name gave me pause while reading the new code. maybe call it `pending_work`?

**Review by danhhz — APPROVED** [review]
persist_source change LGTM! I didn't look at the buffer.rs change. sorry this has been such a saga!

**Review by teskje — APPROVED** [review]
LGTM

**Review by vmarcos — APPROVED** [review]
Took another look at `give_at`; it was actually fine, so this LGTM!

**Pull request merged**
